gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+		gt_fib = &ltbl->fib_tbl[fib_id];
+	else
+		gt_fib = &ltbl->fib_tbl6[fib_id];
+
+	gt_fib->action = GK_FWD_GRANTOR;
+	rte_memcpy(&gt_fib->u.grantor.gt_addr,
+		gt_addr, sizeof(gt_fib->u.grantor.gt_addr));
+	gt_fib->u.grantor.eth_cache = eth_cache;
+
+	ret = lpm_add_route(&prefix->addr, prefix->len, fib_id, ltbl);
+	if (ret < 0) {
+		initialize_fib_entry(gt_fib);
+		return NULL;
+	}
+
+	return gt_fib;
+}
+
+static struct gk_fib *
+init_drop_fib_locked(struct ip_prefix *ip_prefix, struct gk_config *gk_conf)
+{
+	int ret;
+	struct gk_fib *ip_prefix_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	/* Initialize the fib entry for the IP prefix. */
+	int fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
+	if (fib_id < 0)
+		return NULL;
+
+	if (ip_prefix->addr.proto == ETHER_TYPE_IPv4)
+		ip_prefix_fib = &ltbl->fib_tbl[fib_id];
+	else if (likely(ip_prefix->addr.proto == ETHER_TYPE_IPv6))
+		ip_prefix_fib = &ltbl->fib_tbl6[fib_id];
+	else
+		rte_panic("Unexpected condition at gk: unknown IP type %hu at %s",
+			ip_prefix->addr.proto, __func__);
+
+	ip_prefix_fib->action = GK_DROP;
+
+	ret = lpm_add_route(&ip_prefix->addr, ip_prefix->len, fib_id, ltbl);
+	if (ret < 0) {
+		initialize_fib_entry(ip_prefix_fib);
+		return NULL;
+	}
+
+	return ip_prefix_fib;
+}
+
+static int
+add_fib_entry_locked(struct ip_prefix *prefix,
+	struct ipaddr *gt_addr, struct ipaddr *gw_addr,
+	enum gk_fib_action action, struct gk_config *gk_conf)
+{
+	switch (action) {
+	case GK_FWD_GRANTOR: {
+		struct gk_fib *gt_fib;
+
+ 		if (gt_addr == NULL || gw_addr == NULL)
+			return -1;
+
+		gt_fib = init_grantor_fib_locked(
+			prefix, gt_addr, gw_addr, gk_conf);
+		if (gt_fib == NULL)
+			return -1;
+
+		/*
+	 	 * TODO The nexthop MAC address should be initialized.
+	 	 */
+		break;
+	}
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET: {
+		struct gk_fib *gw_fib;
+
+ 		if (gw_addr == NULL)
+			return -1;
+
+		gw_fib = init_gateway_fib_locked(
+			prefix, action, gw_addr, gk_conf);
+		if (gw_fib == NULL)
+			return -1;
+
+		/*
+	 	 * TODO The nexthop MAC address should be initialized.
+	 	 */
+		break;
+	}
+
+	case GK_DROP: {
+		struct gk_fib *ip_prefix_fib =
+			init_drop_fib_locked(prefix, gk_conf);
+		if (ip_prefix_fib == NULL)
+			return -1;
+
+		break;
+	}
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"Invalid fib action %u at %s\n", action, __func__);
+		return -1;
+		break;
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: add a fib entry for ip prefix = %s with action = %u]\n",
+		prefix->str, action);
+
+	return 0;
+}
+
+static int
+check_gateway_prefix(struct ip_prefix *prefix, struct ipaddr *gw_addr)
+{
+	if (gw_addr->proto == ETHER_TYPE_IPv4) {
+		uint32_t ip4_mask =
+			rte_cpu_to_be_32(~0ULL << (32 - prefix->len));
+		if ((prefix->addr.ip.v4.s_addr & ip4_mask) ==
+				(gw_addr->ip.v4.s_addr & ip4_mask))
```

Comment:
2. Shouldn't the above expression return ``-1``? Since if ``gw_addr`` is included in ``prefix``, then ``(prefix->addr.ip.v4.s_addr ^ gw_addr->ip.v4.s_addr) & ip4_mask`` would return ``0``.